airflow/www/views.py

```diff
@@ -4819,8 +4835,21 @@ def process_form(self, form, is_created):
                     # value isn't an empty string.
                     if value != "":
                         extra[field_name] = value
-
         if extra.keys():
+            sensitive_unchanged_keys = set()
+            for key, value in extra.items():
+                if value == SENSITIVE_FIELD_PLACEHOLDER:
+                    sensitive_unchanged_keys.add(key)
+            if sensitive_unchanged_keys:
+                try:
+                    conn = BaseHook.get_connection(conn_id)
+                except AirflowNotFoundException:
+                    conn = None
+                for key in sensitive_unchanged_keys:
+                    if conn and conn.extra_dejson.get(key):
+                        extra[key] = conn.extra_dejson.get(key)
+                    else:
+                        del extra[key]
```

Comment:
```suggestion
                def _iter_masked_extra_items():
                    for key, val in extra.items():
                        if key not in sensitive_unchanged_keys:
                            yield key, val
                        elif conn and (conn_val := conn.extra_dejson.get(key)):
                            yield key, conn_val

                extra = dict(_iter_masked_extra_items())
```

Not entirely sure I got this one right

Reply:
I'd hesitate with getting that one in. It does not bring much value speed/optimisation wise and I think it is slightly less readable.